src/Driver/Http2Driver.php

```diff
@@ -706,27 +707,37 @@ private function parser(string $settings = null): \Generator
             }
 
             while (true) {
-                if (++$frameCount === $maxFramesPerSecond) {
-                    $now = $this->timeReference->getCurrentTime();
-                    if ($lastReset === $now && $bytesReceived / $frameCount < $minAverageFrameSize) {
+                $now = $this->timeReference->getCurrentTime();
+                if ($lastReset === $now) {
+                    // Inspired by nginx flood detection:
+                    // https://github.com/nginx/nginx/commit/af0e284b967d0ecff1abcdce6558ed4635e3e757
+                    if ($totalBytesReceivedSinceReset > $payloadBytesReceivedSinceReset + 8192) {
                         throw new Http2ConnectionException(
                             $this->client,
-                            "Average frame size too low",
+                            "Flood detected",
                             self::ENHANCE_YOUR_CALM
                         );
                     }
-
+                } else {
                     $lastReset = $now;
-                    $frameCount = 0;
-                    $bytesReceived = 0;
+                    $totalBytesReceivedSinceReset = 0;
+                    $payloadBytesReceivedSinceReset = 0;
+                }
+
+                if (empty($this->streams) && $lastStreamOpening < $now - $timeout) {
+                    throw new Http2ConnectionException(
+                        $this->client,
+                        "Too much time elapsed with non-payload frames",
+                        self::ENHANCE_YOUR_CALM
+                    );
```

Comment:
Does this send the unprocessed stream ID to the client?

Reply:
Yes, the shutdown handler will send the highest stream ID processed.

Reply:
Not sure whether self::ENHANCE_YOUR_CALM is the correct error code here?